web3.js/src/connection.ts

```diff
@@ -2673,6 +2676,7 @@ const COMMON_HTTP_HEADERS = {
  */
 export class Connection {
   /** @internal */ _commitment?: Commitment;
+  /** @internal */ _cluster?: Cluster;
```

Comment:
I really feel strongly that these should be a pair.

```typescript
_rpcEndpoint: string | ClusterfulEndpoint;
```

Where…

```typescript
type ClusterfulEndpoint = Readonly<{
  cluster: Cluster,
  url: string,
}>;
```

Then people can continue to supply a string _or_ they can start doing this:

```typescript
new Connection({cluster: 'devnet', url: 'https://coolcustomrpc.com'});
```

Or hell maybe even:

```typescript
new Connection(devnet('https://coolcustomrpc.com'));
```

…where:

```typescript
function devnet(url: string) {
  return {
    cluster: 'devnet',
    url,
  };
}
```

Reply:
Hell we could _even_ use [tagged template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates) if we wanted to waste everyone's time learning something new.

```typescript
new Connection(devnet`https://coolcustomrpc.com`);
```

Reply:
I can see an argument for this -- the two values are semantically entwined, so should be provided together.

This makes passing options clumsy:

```ts
new Connection({ cluster: 'devnet', url: 'https://coolcustomrpc.com' }, 'confirmed')
new Connection({ cluster: 'devnet', url: 'https://coolcustomrpc.com' }, { commitment: 'confirmed', ... })
```

I don't hate `new Connection(devnet('https://coolcustomrpc.com'))` though.

Reply:
I'm just really uncomfortable with `cluster` being so far away from `endpoint`.

Reply:
I'm in favor of having a `ClusterEndpoint` type as well for this, agreed that having helper methods for each cluster helps with api ergonomics